File: Daily Challanges/2026-04-April/2026-04-17.js
function decode(message) {
  const KEY = "VLHCGMDLNH";
  let decoded = [];
  let keyIndex = 0;

  for (let i = 0; i < message.length; i++) {
    let char = message[i];

    if (/[A-Z]/.test(char)) {
      let shiftSize =
        KEY.charCodeAt(keyIndex % KEY.length) - "A".charCodeAt(0) + 1;

      // In JS, % can return negative numbers, so we add 26 before the final modulo
      let charCode = char.charCodeAt(0) - "A".charCodeAt(0);
      let decodedVal = (charCode - shiftSize) % 26;
      if (decodedVal < 0) decodedVal += 26;

      decoded.push(String.fromCharCode(decodedVal + "A".charCodeAt(0)));
      keyIndex++;
    } else {
      decoded.push(char);
    }
  }

  return decoded.join("");
}

console.log(decode("YAVJYNXE"));
console.log(decode("YALLUT PQUMJP"));
console.log(decode("UAC DYR EISAKYM"));
console.log(decode("GQMS NBMZU"));
console.log(decode("W IQQURV UG I ZDMDTRV IVW JQDHY TMHSA QB"));

File: Daily Challanges/2026-04-April/2026-04-17.py
def decode(message):
    n = len(message)
    KEY = "VLHCGMDLNH"
    decoded = []
    key_index = 0
    for i in range(n):
        char = message[i]
        if char.isalpha():
            shift_size = ord(KEY[key_index % len(KEY)]) - ord('A') + 1
            decoded_val = (ord(char) - ord('A') - shift_size) % 26
            decoded.append(chr(decoded_val + ord('A')))
            key_index += 1
        else:
            decoded.append(char)
            
    return "".join(decoded)

print(decode("YAVJYNXE"))
print(decode("YALLUT PQUMJP"))
print(decode("UAC DYR EISAKYM"))
print(decode("GQMS NBMZU"))
print(decode("W IQQURV UG I ZDMDTRV IVW JQDHY TMHSA QB"))